flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeSubpartitionReader.java

```diff
@@ -176,21 +202,28 @@ public void resumeConsumption() {
 
     @Override
     public Throwable getFailureCause() {
-        // we can never throw an error after this was created
-        return null;
+        synchronized (lock) {
+            return failureCause;
+        }
     }
 
     @Override
     public boolean isAvailable(int numCreditsAvailable) {
-        if (numCreditsAvailable > 0) {
-            return !buffersRead.isEmpty();
-        }
+        synchronized (lock) {
+            if (isReleased) {
+                return true;
+            }
 
-        return !buffersRead.isEmpty() && !buffersRead.peek().isBuffer();
+            if (numCreditsAvailable > 0) {
```

Comment:
Perhaps change to 

```
return !buffersRead.isEmpty() && (numCredit > 0 || !bufferReader.peek().isBuffer())
```

or 

```

if (buffersRead.isEmpty()) {
      return false;
}

return numCreditsAvailable > 0 || !buffersRead.peek().isBuffer();
```

to avoid duplicate condition ?